ros_babel_fish/src/message.cpp

```diff
@@ -30,104 +30,101 @@ Message::~Message() = default;
+constexpr inBounds( const T &val )
 {
-  return std::numeric_limits<U>::min() <= val &&
-         val <= std::numeric_limits<U>::max();
+  return std::numeric_limits<U>::min() <= val && val <= std::numeric_limits<U>::max();
 }
 
 template<typename T, typename U>
-typename std::enable_if<std::is_integral<U>::value, bool>::type isCompatible( const T &val )
+typename std::enable_if<std::is_floating_point<T>::value, bool>::type
+constexpr inBounds( const T &val )
 {
-  if ( std::is_same<T, U>::value ) return true;
-  if ( std::is_floating_point<T>::value ) return false;
-  return inBounds<T, U>( val );
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wimplicit-int-float-conversion"
+  return std::numeric_limits<U>::min() <= val && val <= std::numeric_limits<U>::max();
+#pragma GCC diagnostic pop
 }
 
+/**
+ * Returns true if type T can always be stored in U without truncation or loss of precision.
+ */
 template<typename T, typename U>
-typename std::enable_if<std::is_floating_point<U>::value, bool>::type isCompatible( const T & )
+typename std::enable_if<std::is_arithmetic<T>::value && std::is_arithmetic<U>::value, bool>::type
+constexpr isCompatible()
 {
-  return std::is_integral<T>::value || sizeof( T ) <= sizeof( U );
+  // See https://en.cppreference.com/w/cpp/types/numeric_limits/digits
+  // + 1 added because ::digits returns nbits-1 for signed types.
+  return std::is_same<rm_cvref_t<T>, rm_cvref_t<U>>::value ||
+         ((std::is_integral<T>::value || std::is_floating_point<U>::value) &&
+          (std::numeric_limits<T>::digits + 1 < std::numeric_limits<U>::digits));
 }
 
 template<typename T, typename U>
 void assignValue( Message *m, const T &value )
 {
   using namespace message_type_traits;
-  if ( m->type() != message_type<T>::value && !isCompatible<T, U>( value ))
-    throw BabelFishException(
-      "Value does not fit into value message! Make sure you're using the correct type or at least stay within the range of values for the message type!" );
+  if ( !isCompatible<T, U>() )
+  {
+    if ( !inBounds<T, U>( value ) )
+    {
+      throw BabelFishException(
+          "Value does not fit into value message! Make sure you're using the correct type or at least stay within the range of values for the message type!");
+    }
 #if RBF_WARN_ON_INCOMPATIBLE_TYPE
-  if ( m->type() != message_type<T>::value && !isCompatible<T, U>())
     ROS_WARN_ONCE_NAMED( "RosBabelFish",
                          "Assigned value fits but the type of the assignment can not be converted without loss of information in some cases! This message is printed only once!" );
 #endif
-  m->as<ValueMessage<U>>().setValue( static_cast<U>(value));
+  }
+  m->as<ValueMessage<U>>().setValue( static_cast<U>(value) );
 }
 
 template<typename T>
 void assignToValueMessage( Message *m, const T &value )
 {
   using namespace message_type_traits;
-  switch ( m->type())
+  switch ( m->type() )
```

Comment:
The code style doesn't use a space between parentheses.
If you are using CLion, you can use the attached zipped code style.
[ROS.zip](https://github.com/StefanFabian/ros_babel_fish/files/7152586/ROS.zip)

Reply:
Thanks, I was indeed wondering how these should actually be formatted. I would highly recommend adding a `.clang-format` to the repo, which takes out the guesswork, but I understand that the current style is not well supported by it.

Reply:
Yes, I was thinking about adding one but I would have to check what is supported. I think in the past it wasn't really the same and it was a bit cumbersome to get CLion to use it.

Reply:
Clang-format support has been very seamless in CLion for a while now and it apparently supports exporting `.clang-format` directly from its formatting config as well. I also experimented with https://github.com/valgur/clang-guess-format but it fails due to clang-format missing an option to avoid adding a space between adjacent parentheses like CLion does by default.